pkg/bitwarden/cli.go

```diff
@@ -0,0 +1,153 @@
+	client := bitwardenCliClient{
+		username: username,
+		password: password,
+		run:      run,
+	}
+	return &client, client.loginAndListItems()
+}
+
+type bwLoginResponse struct {
+	Success bool `json:"success"`
+	Data    struct {
+		Raw string `json:"raw"`
+	} `json:"data"`
+}
+
+func (c *bitwardenCliClient) loginAndListItems() error {
+	c.Lock()
+	defer c.Unlock()
+	output, err := c.runBWCmd(bwLoginCmd, c.username, c.password, "--response")
+	if err != nil {
+		return err
+	}
+	r := bwLoginResponse{}
+	if err := json.Unmarshal(output, &r); err != nil {
+		//TODO need to censor if we plan this on a job
+		return err
+	}
+	if r.Success {
+		raw := r.Data.Raw
+		if raw != "" {
+			c.session = raw
+			if len(c.savedItems) > 0 {
+				return nil
+			}
+			var items []Item
+			out, err := c.runBWCmd(bwListCmd, "items")
+			if err != nil {
+				return err
+			}
+			err = json.Unmarshal(out, &items)
+			if err != nil {
+				return err
+			}
+			c.savedItems = items
+			return nil
+		}
+		// should never happen
+		return fmt.Errorf("bw-login succeeded with empty '.data.raw'")
+	}
+	// should never happen
+	return fmt.Errorf("bw-login failed without run's error")
+}
+
+func (c *bitwardenCliClient) GetItem(itemName string) (ItemGetter, error) {
+	var itemGetter cliItemGetter
+	for _, i := range c.savedItems {
+		if i.Name == itemName {
+			itemGetter.item = i
+			itemGetter.client = c
+			return &itemGetter, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to find item with name %s", itemName)
+}
+
+func (c *bitwardenCliClient) Logout() error {
+	c.Lock()
+	defer c.Unlock()
+	if _, err := c.runBWCmd(bwLogoutCmd, "--response"); err != nil {
+		return err
+	}
+	c.session = ""
+	c.savedItems = []Item{}
+	return nil
+}
+
+type cliItemGetter struct {
+	client *bitwardenCliClient
+	item   Item
+}
+
+func (getter cliItemGetter) GetField(filedName string) (string, error) {
+	for _, field := range getter.item.Fields {
+		if field.Name == filedName {
+			return field.Value, nil
+		}
+	}
+	return "", fmt.Errorf("failed to find field %s in item %s", filedName, getter.item.Name)
+}
+
+func (getter cliItemGetter) GetAttachment(attachmentName string) ([]byte, error) {
+	for _, attachment := range getter.item.Attachments {
+		if attachment.FileName == attachmentName {
+			//bw --session "${BW_SESSION}" get attachment "${attachment_id}" --itemid "${item_id}" --raw
+			out, err := getter.client.runBWCmd(bwGetCmd, "attachment", attachment.ID, "--itemid", getter.item.ID, "--raw")
+			return out, err
+		}
+	}
+	return nil, fmt.Errorf("failed to find attachment %s in item %s", attachmentName, getter.item.Name)
+}
+
+type bkCmdType string
+
+const (
+	bwLoginCmd  bkCmdType = "login"
+	bwLogoutCmd bkCmdType = "logout"
+	bwListCmd   bkCmdType = "list"
+	bwGetCmd    bkCmdType = "get"
+)
+
+func (c *bitwardenCliClient) runBWCmd(cmdType bkCmdType, args ...string) ([]byte, error) {
```

Comment:
This still seems off. Why isn't there `bitwardenCliClient.GetAttachment` ? Why is it the `cliItemGetter` that knows how to use the `cliClient` to get an attachment?

Why don't you refactor this so the client (that exposes `GetAttachment` and `GetItem` etc) is in a separate package from the CLI implementation, that should help you separate concerns.

Reply:
The function is deleted.